cpp/include/cudf/detail/utilities/dremel.hpp

```diff
@@ -0,0 +1,195 @@
+ * of the leaf level. The values for the merge are the definition and repetition levels
+ * ```
+ * empties at level 1 = {0, 5}
+ * def values at 1    = {1, 1}
+ * rep values at 1    = {1, 1}
+ * indices at leaf    = {0, 1, 2, 3, 4}
+ * def values at leaf = {2, 2, 2, 2, 2}
+ * rep values at leaf = {2, 2, 2, 2, 2}
+ * ```
+ *
+ * merged def values  = {1, 2, 2, 2, 2, 2, 1}
+ * merged rep values  = {1, 2, 2, 2, 2, 2, 1}
+ *
+ * The size of the rep/def values is now larger than the leaf values and the offsets need to be
+ * adjusted in order to point to the correct start indices. We do this with an exclusive scan over
+ * the indices of offsets of empty lists and adding to existing offsets.
+ * ```
+ * Level 1 new offsets = {0, 1, 4, 6, 7}
+ * ```
+ * Repetition values at the beginning of a list need to be decremented. We use the new offsets to
+ * scatter the rep value.
+ * ```
+ * merged rep values  = {1, 2, 2, 2, 2, 2, 1}
+ * scatter (1, new offsets)
+ * new offsets        = {0, 1,       4,    6, 7}
+ * new rep values     = {1, 1, 2, 2, 1, 2, 1}
+ * ```
+ *
+ * Similarly we merge up all the way till level 0 offsets
+ *
+ * STRUCT COLUMNS :
+ * In case of struct columns, we don't have to merge struct levels with their children because a
+ * struct is the same size as its children. e.g. for a column `struct<int, float>`, if the row `i`
+ * is null, then the children columns `int` and `float` are also null at `i`. They also have the
+ * null entry represented in their respective null masks. So for any case of strictly struct based
+ * nesting, we can get the definition levels merely by iterating over the nesting for the same row.
+ *
+ * In case struct and lists are intermixed, the definition levels of all the contiguous struct
+ * levels can be constructed using the aforementioned iterative method. Only when we reach a list
+ * level, we need to do a merge with the subsequent level.
+ *
+ * So, for a column like `struct<list<int>>`, we are going to merge between the levels `struct<list`
+ * and `int`.
+ * For a column like `list<struct<int>>`, we are going to merge between `list` and `struct<int>`.
+ *
+ * In general, one nesting level is the list level and any struct level that precedes it.
+ *
+ * A few more examples to visualize the partitioning of column hierarchy into nesting levels:
+ * (L is list, S is struct, i is integer(leaf data level), angle brackets omitted)
+ * ```
+ * 1. LSi     = L   Si
+ *              - | --
+ *
+ * 2. LLSi    = L   L   Si
+ *              - | - | --
+ *
+ * 3. SSLi    = SSL   i
+ *              --- | -
+ *
+ * 4. LLSLSSi = L   L   SL   SSi
+ *              - | - | -- | ---
+ * ```
+ *
+ * @param col Column of LIST type
+ * @param level_nullability Pre-determined nullability at each list level. Empty means infer from
+ * `col`
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ *
+ * @return A struct containing dremel data
+ */
+dremel_data get_dremel_data(column_view h_col,
```

Comment:
Should this require a `lists_column_view`?

Reply:
Right now this function assumes that we have a `column_view` of list dtype. It should probably validate that the input column's dtype actually is list, and passing a `lists_column_view` would be one way to promise that. It definitely requires some minor changes to the rest of the code, but I think it could work. I'd say that's also a change for the next PR, but definitely worth investigating.